### PR TITLE
Extensions examples

### LINKED ISSUE #505: end-to-end tutorial
### Problem/Use Case
Write a tutorial that explains how to:
    1. define an extension
    2. define a Container class for the extension
    3. write instances of the Container class 
    4. read the Container class(es) instances back in

## Checklist

- [ ] Have you ensured the feature or change was not already [reported](https://github.com/NeurodataWithoutBorders/pynwb/issues/) ?
- [ ] Have you included a brief and descriptive title?
- [ ] Have you included a clear description of the problem you are trying to solve?
- [ ] Have you included a minimal code snippet that reproduces the issue you are encountering?

### PR DESCRIPTION
**Pull request by bendichter**
add cortical surface example to Extensions tutorial

fix #505

### CODE REVIEW

File: docs/gallery/general/extensions.py [hunk 1]
```diff
@@ -124,3 +124,113 @@ def __init__(self, *args, **kwargs):
 #     When defining your own :py:class:`~pynwb.core.NWBContainer`, the subclass name does not need to be the same
 #     as the extension type name. However, it is encouraged to keep class and extension names the same for the
 #     purposes of readibility.
+
+
+####################
+# Example: Cortical Surface Mesh
+# -----------------------------------------------------
+#
+# Here we show how to create extensions by example by creating a data class for a
```
Comment (nicain) [inline docs/gallery/general/extensions.py:133]: typo: "by example by"

File: docs/gallery/general/extensions.py [hunk 2]
```diff
@@ -124,3 +124,113 @@ def __init__(self, *args, **kwargs):
...
+# cortical surface mesh. This data type is particularly important for ECoG data,
+# we need to know where each electrode is with respect to the gyri and sucli.
```
Comment (nicain) [inline docs/gallery/general/extensions.py]: change "we need" to "because we need"?

File: docs/gallery/general/extensions.py [hunk 3]
```diff
@@ -124,3 +124,113 @@ def __init__(self, *args, **kwargs):
...
+# Surface mesh objects contain two types of data:
+#
+# 1. `vertices`, which is an (n, 3) matrix of floats that represents points in 3D space.
+#
+# 2. `faces`, which is an (n, 3) matrix of uints that represents indices of the `vertices` matrix. Each triplet of
+# points defines a triangular face, and the mesh is comprised of a collection of triangular faces.
```
Comment (nicain) [inline docs/gallery/general/extensions.py:141]: Reminds me of this: http://paulbourke.net/dataformats/ply/

File: docs/gallery/general/extensions.py [hunk 4]
```diff
@@ -124,3 +124,113 @@ def __init__(self, *args, **kwargs):
...
+
+from pynwb.spec import NWBDatasetSpec, NWBNamespaceBuilder, NWBGroupSpec
+
+name = 'ecog'
+ns_path = name + ".namespace.yaml"
+ext_source = name + ".extensions.yaml"
+
+####################
+# Now we define the data structures. We use :py:meth:`~pynwb.core.NWBDataInterface` as the base type,
+# which is the lowest type you are likely to use as a base. The name of the
```
Comment (nicain) [inline docs/gallery/general/extensions.py:153]: "lowest"; what does this mean?  Least complicated?

File: docs/gallery/general/extensions.py [hunk 5]
```diff
@@ -124,3 +124,113 @@ def __init__(self, *args, **kwargs):
...
+                           shape=(None, 3),
+                           name='vertices', dtype='float',
+                           dims=('vertex_number', 'xyz'))]
+
+surface = NWBGroupSpec(doc='brain cortical surface',
+                       datasets=datasets,
+                       neurodata_type_def='CorticalSurface',
+                       neurodata_type_inc='NWBDataInterface')
+
+# Now we set up the builder and add this object
```
Comment (nicain) [inline docs/gallery/general/extensions.py:171]: add a "#######" above if you want sphinx-gallery to separate, as opposed to inline code comment

File: docs/gallery/general/extensions.py [hunk 6]
```diff
@@ -124,3 +124,113 @@ def __init__(self, *args, **kwargs):
...
+
+ns_builder = NWBNamespaceBuilder(name + ' extensions', name)
+ns_builder.add_spec(ext_source, surface)
+ns_builder.export(ns_path)
+
+####################
+# The above should generate 2 YAML files. `ecog.extensions.yaml`,
+# defines the newly defined types
+#
```
Comment (nicain) [inline docs/gallery/general/extensions.py:180]: Not sure how this will typeset; maybe use the prototype tool, just to make sure it looks how you want

**Review by nicain — CHANGES_REQUESTED** [review]

**Comment by oruebel** [comment]
@bendichter  What is the status of this PR. Looks like there are a few simple typos but otherwise good to go.

File: docs/gallery/general/extensions.py (hunk 1 above)
Comment (bendichter) [reply docs/gallery/general/extensions.py:133]:
```suggestion
# Here we show how to create extensions by creating a data class for a
```

**Comment by bendichter** [comment]
closed in favor of #1040

**Pull request closed without merging**